Improve behavior when testing without an internet connection

TaylorSMarks opened the pull request:
I often work on my django project from my laptop when it lacks an internet connection. The django-bootstrap-icons library does not gracefully handle that... there's an uncaught ConnectionError raised by requests which causes the entire page to fail to load.

This PR makes it so that that ConnectionError is caught, permitting the page to otherwise load. It also makes it so that an alternate message can be provided from the template, so that something other than a long message about how the icon failed to load is rendered.

Personally, I like using "!!!" as my alt... it takes up about as much space as the icon that would otherwise be loaded. It can also be styled, by using something like:

    {% bs_icon 'bell' color='white' size='2em' extra_classes='bell' alt='<span style="color:red">!!!</span>' %}
    
Not sure whether the ability to use raw html as part of the alt string should be considered a vulnerability or not... it doesn't really enable anything that couldn't have been done anyways via the initial name argument, given that gets rendered in the HTML without being escaped, too.

TaylorSMarks committed:
Improve behavior when testing without an internet connection

TaylorSMarks commented:
I'm not sure how you want the imports list ordered. It can be changed however.

I'm not a fan of pylint and would just ignore its complaint that the icon function can take up to six arguments.

christianwgd commented:
First, you're right, this clearly can be improved. I never had this problem, since I'm using some other libraries from CDNs and in most cases the whole project wouldn't run. But I see your point...

Instead of using an additional parameter, I would prefer to add a new setting, so you can customize the output you get, if your machine is offline. This can be a string (that defaults to the previous behaviour) like your '!!!' or the svg code of an "no-icon" icon. I've changed and uploaded the code, let me know what you think...

By the way: Did you notice the new feature that I added, the icon cache? With the cache enabled you can run your app once and get all icons in the cache. Now you can go offline and get all icons from the cache.

TaylorSMarks commented:
I saw the cache code - it didn’t occur to me until now that the cache is written to file and so persisted between runs.

That cache probably renders my changes unnecessary. I’ll try it out - you can probably reject my PR.

Taylor

Sent from my iPhone

> On Jun 12, 2021, at 15:42, Christian Wiegand ***@***.***> wrote:
> 
> ﻿
> First, you're right, this clearly can be improved. I never had this problem, since I'm using some other libraries from CDNs and in most cases the whole project wouldn't run. But I see your point...
> 
> Instead of using an additional parameter, I would prefer to add a new setting, so you can customize the output you get, if your machine is offline. This can be a string (that defaults to the previous behaviour) like your '!!!' or the svg code of an "no-icon" icon. I've changed and uploaded the code, let me know what you think...
> 
> By the way: Did you notice the new feature that I added, the icon cache? With the cache enabled you can run your app once and get all icons in the cache. Now you can go offline and get all icons from the cache.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
